internal/runtime/client/test/fake_client.go

```diff
@@ -0,0 +1,149 @@
+}
+
+func NewFakeRuntimeClientBuilder() *FakeRuntimeClientBuilder {
+	return &FakeRuntimeClientBuilder{}
+}
+
+func (f *FakeRuntimeClientBuilder) WithCatalog(catalog *runtimecatalog.Catalog) *FakeRuntimeClientBuilder {
+	f.catalog = catalog
+	return nil
+}
+
+func (f *FakeRuntimeClientBuilder) WithCallAllExtensionResponses(responses map[runtimecatalog.GroupVersionHook]runtimehooksv1.ResponseObject) *FakeRuntimeClientBuilder {
+	f.callAllResponses = responses
+	return f
+}
+
+func (f *FakeRuntimeClientBuilder) WithCallExtensionResponses(responses map[string]runtimehooksv1.ResponseObject) *FakeRuntimeClientBuilder {
+	f.callResponses = responses
+	return f
+}
+
+func (f *FakeRuntimeClientBuilder) MarkReady(ready bool) *FakeRuntimeClientBuilder {
+	f.ready = ready
+	return f
+}
+
+func (f *FakeRuntimeClientBuilder) Build() runtimeclient.Client {
+	if f.catalog == nil {
+		f.catalog = DefaultTestCatalog
+	}
+	return &fakeRuntimeClient{
+		isReady:          f.ready,
+		callAllResponses: f.callAllResponses,
+		callResponses:    f.callResponses,
+		catalog:          f.catalog,
+	}
+}
+
+var _ runtimeclient.Client = &fakeRuntimeClient{}
+
+type fakeRuntimeClient struct {
+	isReady          bool
+	catalog          *runtimecatalog.Catalog
+	callAllResponses map[runtimecatalog.GroupVersionHook]runtimehooksv1.ResponseObject
+	callResponses    map[string]runtimehooksv1.ResponseObject
+}
+
+// CallAllExtensions implements Client.
+func (fc *fakeRuntimeClient) CallAllExtensions(ctx context.Context, hook runtimecatalog.Hook, request runtime.Object, response runtimehooksv1.ResponseObject) error {
+	gvh, err := fc.catalog.GroupVersionHook(hook)
+	if err != nil {
+		return errors.Wrap(err, "failed to compute GVH")
+	}
+	expectedResponse, ok := fc.callAllResponses[gvh]
+	if !ok {
+		// This should actually panic because an error here would mean a mistake in the test setup.
+		panic(fmt.Sprintf("test response not available hook for %q", gvh))
+	}
+	if err := fc.catalog.Convert(expectedResponse, response, ctx); err != nil {
+		// This should actually panic because an error here would mean a mistake in the test setup.
+		panic("cannot update response")
+	}
+	if response.GetStatus() == runtimehooksv1.ResponseStatusFailure {
+		return errors.Errorf("runtime hook %q failed", gvh)
+	}
+	return nil
+}
+
+// CallExtension implements Client.
+func (fc *fakeRuntimeClient) CallExtension(ctx context.Context, _ runtimecatalog.Hook, name string, request runtime.Object, response runtimehooksv1.ResponseObject) error {
+	expectedResponse, ok := fc.callResponses[name]
+	if !ok {
+		// This should actually panic because an error here would mean a mistake in the test setup.
+		panic(fmt.Sprintf("test response not available for extension %q", name))
+	}
+	if err := fc.catalog.Convert(expectedResponse, response, ctx); err != nil {
+		// This should actually panic because an error here would mean a mistake in the test setup.
+		panic("cannot update response")
+	}
+	return nil
```

Comment:
```
	if response.GetStatus() == runtimehooksv1.ResponseStatusFailure {
		return errors.Errorf("ExtensionHandler %s failed with message %s", name, response.GetMessage())
	}

	return nil
```
(copied from regular client, otherwise the behavior for failure responses diverges)